catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/OperationsStorageSupport.java

```diff
@@ -0,0 +1,147 @@
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.impl.operations;
+
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.Constants;
+import ddf.catalog.content.StorageException;
+import ddf.catalog.content.StorageProvider;
+import ddf.catalog.content.data.ContentItem;
+import ddf.catalog.content.operation.StorageRequest;
+import ddf.catalog.history.Historian;
+import ddf.catalog.source.IngestException;
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.util.impl.Requests;
+
+public class OperationsStorageSupport {
+    private static final Logger LOGGER = LoggerFactory.getLogger(OperationsStorageSupport.class);
+
+    private static final Logger INGEST_LOGGER =
+            LoggerFactory.getLogger(Constants.INGEST_LOGGER_NAME);
+
+    //
+    // Injected propertiers
+    //
+    private final SourceOperations sourceOperations;
+
+    private final QueryOperations queryOperations;
+
+    private Historian historian;
+
+    public OperationsStorageSupport(SourceOperations sourceOperations,
+            QueryOperations queryOperations) {
+        this.sourceOperations = sourceOperations;
+        this.queryOperations = queryOperations;
+    }
+
+    public void setHistorian(Historian historian) {
+        this.historian = historian;
+    }
+
+    <T extends StorageRequest> T prepareStorageRequest(T storageRequest,
+            Supplier<List<ContentItem>> getContentItems)
+            throws IngestException, SourceUnavailableException {
+        validateStorageRequest(storageRequest, getContentItems);
+
+        queryOperations.setFlagsOnRequest(storageRequest);
+
+        if (Requests.isLocal(storageRequest) && (!sourceOperations.isSourceAvailable(
+                sourceOperations.getCatalog())
+                || !isStorageAvailable(sourceOperations.getStorage()))) {
+            String message = "Local provider is not available, cannot perform storage operation.";
+            SourceUnavailableException sourceUnavailableException = new SourceUnavailableException(
+                    message);
+            if (INGEST_LOGGER.isWarnEnabled()) {
+                INGEST_LOGGER.warn(message, sourceUnavailableException);
+            }
+            throw sourceUnavailableException;
+        }
+
+        return storageRequest;
+    }
+
+    void commitAndCleanup(StorageRequest storageRequest, Optional<String> historianTransactionKey,
+            HashMap<String, Path> tmpContentPaths) {
+        if (storageRequest != null) {
+            try {
+                sourceOperations.getStorage()
+                        .commit(storageRequest);
+                historianTransactionKey.ifPresent(historian::commit);
```

Comment:
Might be able to do that. I'm not familiar with the internals of the Historian well enough to say if that would be an improvement.